hoomd/md/FrictionPair.h

```diff
@@ -0,0 +1,757 @@
+
+        uint16_t seed = this->m_sysdef->getSeed();
+
+        // for each particle
+        for (int i = 0; i < (int)m_pdata->getN(); i++)
+            {
+            // access the particle's position and type (MEM TRANSFER: 4 scalars)
+            Scalar3 pi = make_scalar3(h_pos.data[i].x, h_pos.data[i].y, h_pos.data[i].z);
+
+            quat<Scalar> quat_i(h_orientation.data[i]);
+
+            // get the velocitie of i-th particle and transform it from body to principle frame
+            Scalar3 vel_i = make_scalar3(h_vel.data[i].x, h_vel.data[i].y, h_vel.data[i].z);
+
+            unsigned int typei = __scalar_as_int(h_pos.data[i].w);
+            Scalar dia_i = h_diameter.data[i];
+
+            // calculate angmom_i (angular momentum of th particle in the body frame)
+            quat<Scalar> angmom_i(h_angmom.data[i]);
+            quat<Scalar> qxP_i = conj(quat_i) * angmom_i;
+            vec3<Scalar> bf_vel_i = Scalar(0.5)
+                                    * vec3(qxP_i.v.x / h_moment_inertia.data[i].x,
+                                           qxP_i.v.y / h_moment_inertia.data[i].y,
+                                           qxP_i.v.z / h_moment_inertia.data[i].z);
+
+            // Rotate angular velocity into global frame
+            vec3<Scalar> gf_angvel_i = rotate(quat_i, bf_vel_i);
+            Scalar3 angvel_i = make_scalar3(gf_angvel_i.x, gf_angvel_i.y, gf_angvel_i.z);
+
+            // sanity check
+            assert(typei < m_pdata->getNTypes());
+
+            // access charge (if needed)
+            Scalar qi = Scalar(0.0);
+            Scalar mass_i = Scalar(0.0);
+            if (friction_evaluator::needsCharge())
+                qi = h_charge.data[i];
+
+            if (friction_evaluator::needsNu())
+                mass_i = h_vel.data[i].w;
+
+            // initialize current particle force, torque, potential energy, and virial to 0
+            Scalar fxi = Scalar(0.0);
+            Scalar fyi = Scalar(0.0);
+            Scalar fzi = Scalar(0.0);
+            Scalar txi = Scalar(0.0);
+            Scalar tyi = Scalar(0.0);
+            Scalar tzi = Scalar(0.0);
+            Scalar pei = Scalar(0.0);
+            Scalar virialxxi = 0.0;
+            Scalar virialxyi = 0.0;
+            Scalar virialxzi = 0.0;
+            Scalar virialyyi = 0.0;
+            Scalar virialyzi = 0.0;
+            Scalar virialzzi = 0.0;
+
+            // loop over all of the neighbors of this particle
+            const size_t myHead = h_head_list.data[i];
+            const unsigned int size = (unsigned int)h_n_neigh.data[i];
+            for (unsigned int k = 0; k < size; k++)
+                {
+                // access the index of this neighbor (MEM TRANSFER: 1 scalar)
+                unsigned int j = h_nlist.data[myHead + k];
+                assert(j < m_pdata->getN() + m_pdata->getNGhosts());
+
+                // calculate dr_ji (MEM TRANSFER: 3 scalars / FLOPS: 3)
```

Comment:
```suggestion
                // calculate dr_ji
```
I don't bother counting memory transfers and flops any more. Any performance is fine as long as the code performs well enough to achieve research goals.